File: app/src/androidTest/java/de/westnordost/streetcomplete/quests/QuestAnswerComponentTest.java
package de.westnordost.streetcomplete.quests;

import androidx.annotation.Nullable;

import org.junit.Test;

import java.util.ArrayList;
import java.util.List;

import de.westnordost.streetcomplete.data.QuestGroup;

import static org.junit.Assert.assertEquals;

public class QuestAnswerComponentTest
{
	@Test public void getSet()
	{
		QuestAnswerComponent c1 = new QuestAnswerComponent();
		c1.onCreate(QuestAnswerComponent.createArguments(11, QuestGroup.OSM));

		assertEquals(QuestGroup.OSM, c1.getQuestGroup());
		assertEquals(11, c1.getQuestId());

		QuestAnswerComponent c2 = new QuestAnswerComponent();
		c2.onCreate(c1.getArguments());

		assertEquals(c2.getQuestGroup(), c1.getQuestGroup());
		assertEquals(c2.getQuestId(), c1.getQuestId());
	}

	@Test public void listener()
	{
		QuestAnswerComponent c1 = new QuestAnswerComponent();

		final int expectQuestId = 3;
		final QuestGroup expectGroup = QuestGroup.OSM_NOTE;
		final String expectNote = "test";
		final String expectQuestTitle = "What?";
		final String expectObject = "jo";
		final ArrayList<String> expectImagePaths = new ArrayList<>();
		expectImagePaths.add("dings");
		expectImagePaths.add("dongs");

		c1.onAttach(new OsmQuestAnswerListener()
		{
			@Override public void onAnsweredQuest(long questId, QuestGroup group, Object answer)
			{
				assertEquals(expectQuestId, questId);
				assertEquals(expectGroup, group);
				assertEquals(expectObject, answer);
			}

			@Override
			public void onComposeNote(long questId, QuestGroup group, String questTitle)
			{
				assertEquals(expectQuestId, questId);
				assertEquals(expectGroup, group);
				assertEquals(expectQuestTitle, questTitle);
			}

			@Override public void onLeaveNote(long questId, QuestGroup group, String questTitle,
											  String note, @Nullable List<String> imagePaths)
			{
				assertEquals(expectQuestId, questId);
				assertEquals(expectGroup, group);
				assertEquals(expectNote, note);
				assertEquals(expectQuestTitle, questTitle);
				assertEquals(expectImagePaths, imagePaths);
			}

			@Override public void onSkippedQuest(long questId, QuestGroup group)
			{
				assertEquals(expectQuestId, questId);
				assertEquals(expectGroup, group);
			}
		});

		c1.onCreate(QuestAnswerComponent.createArguments(expectQuestId, expectGroup));
		c1.onComposeNote(expectQuestTitle);
		c1.onLeaveNote(expectQuestTitle, expectNote, expectImagePaths);
		c1.onAnswerQuest(expectObject);
		c1.onSkippedQuest();
	}
}

File: app/src/androidTest/java/de/westnordost/streetcomplete/quests/QuestAnswerComponentTest.kt
package de.westnordost.streetcomplete.quests

import org.junit.Test

import de.westnordost.streetcomplete.data.QuestGroup

import org.junit.Assert.assertEquals

class QuestAnswerComponentTest {

    @Test fun getSet() {
        val c1 = QuestAnswerComponent()
        c1.onCreate(QuestAnswerComponent.createArguments(11, QuestGroup.OSM))

        assertEquals(QuestGroup.OSM, c1.questGroup)
        assertEquals(11, c1.questId)

        val c2 = QuestAnswerComponent()
        c2.onCreate(c1.arguments)

        assertEquals(c2.questGroup, c1.questGroup)
        assertEquals(c2.questId, c1.questId)
    }

    @Test fun listener() {
        val c1 = QuestAnswerComponent()

        val expectQuestId = 3
        val expectGroup = QuestGroup.OSM
        val expectQuestTitle = "What?"
        val expectObject = "jo"

        c1.onAttach(object : OsmQuestAnswerListener {
            override fun onAnsweredQuest(questId: Long, group: QuestGroup, answer: Any) {
                assertEquals(expectQuestId.toLong(), questId)
                assertEquals(expectGroup, group)
                assertEquals(expectObject, answer)
            }

            override fun onComposeNote(questId: Long, group: QuestGroup, questTitle: String) {
                assertEquals(expectQuestId.toLong(), questId)
                assertEquals(expectGroup, group)
                assertEquals(expectQuestTitle, questTitle)
            }

            override fun onSplitWay(osmQuestId: Long) {
                assertEquals(expectQuestId.toLong(), osmQuestId)
            }

            override fun onSkippedQuest(questId: Long, group: QuestGroup) {
                assertEquals(expectQuestId.toLong(), questId)
                assertEquals(expectGroup, group)
            }
        })

        c1.onCreate(QuestAnswerComponent.createArguments(expectQuestId.toLong(), expectGroup))
        c1.onComposeNote(expectQuestTitle)
        c1.onSplitWay()
        c1.onAnsweredQuest(expectObject)
        c1.onSkippedQuest()
    }
}
